### ISSUE TITLE
Genesis controller not detected on startup, only after ROM reload (Ctrl-R)

### ISSUE THREAD

**Issue by sa666666** [stella-emu/stella#206]
This problem was originally found for Scramble, but was found to exist in Stay Frosty 2 as well.  See here for more details: http://atariage.com/forums/topic/245610-scramble-2600/page-18#entry3830912.

Basically, if the controller is set to 'Genesis' in the ROM properties and the ROM started, the autodetection for the genesis controller doesn't work.  But if you press Ctrl-R while the ROM is running, the autodetection _does_ work.  The same code paths should be followed in both cases, so I don't see how this is happening.

**Comment by sa666666** [comment]
@DirtyHairy, I'm assigning you to this one too, since it seems to be related to the new paddle code that you wrote (somewhat indirectly).  The issue seems to be the following code from Genesis.cxx:
```
  // The Genesis has one more button (C) that can be read by the 2600
  // However, it seems to work opposite to the BoosterGrip controller,
  // in that the logic is inverted
  updateAnalogPin(
    Five,
    (myEvent.get(myFire2Event) == 0) ? minimumResistance : maximumResistance
  );
```
As described in the above thread, the ROM uses the following logic to determine whether the Genesis is plugged in:
```
    lda INPT1           ; Bit 7 will be on if genesis controller is plugged in
    and #$80            ; and button C is NOT pressed.  Bit 7 will be off if
    sta MMcontroller    ; standard joystick is plugged in.
```
However, it looks like in the first block of code that when button C (aka, INPT1) is not pressed (ie, when the event is 0), then INPT1 is set to 0, which according to the second block of code selects the joystick.  So that is one issue.  The second issue is why is works when the ROM is reloaded; it seems to me the autodetection should fail in all cases, according to the current code.

What do you think?

**Comment by ale-79** [comment]
If you start the game (I tried Stay Forsty 2) with the "-debug" option, bit 7 of INPT1 is low (and button C is displayed as pressed), so the ganesis controller is not detected. This get fixed if you exit and reenter the debugger. At this point if you restart the game by using the "reset" command in the prompt and toggle the debugger off again, the controller is detected.
If, instead, you reset with CTRL-R during emulation, bit 7 is low again.

This is different than starting the game in normal mode, as in that case the bug only happens the first time but not after resetting with CTRL-R.

**Comment by sa666666** [comment]
More experimentation.  If I set a break on $F1D6 (the code in the second block above), INPT1 does indeed give 0, and the joystick is detected.  After a time, INPUT prints $80, but it's still at 0 when the breakpoint is first hit.  After a ROM reload, INPT1 is still $80, so the Genesis controller is detected correctly.

The question is, why is INPT1 initially 0, and why does it change to $80 only later??

**Comment by sa666666** [comment]
@ale-79, it looks like we were typing a response at the same time, and yours got through first.  I agree with everything you said except the emulation mode stuff; for me, doing a ROM reload either while in emulation or through the debugger results in the Genesis being detected the second time.  It's just the first time that it fails.  So my original question to @DirtyHairy stands; why is INPT1 0 the first time??

**Comment by ale-79** [comment]
I think that tecnically that should be correct as there's a pullup resistor (10Kohm, if I remember correctly) between pin 5 (INPT1 bit 7) and +5V, so it should take some time for the cap to charge and set INPT1 HI. 
So maybe the problem is that it takes longer than on real hardware and the autodetection misses it?

**Comment by sa666666** [comment]
OK, that makes sense.  In that case, the paddle readout circuitry needs to be modified a little to react a little faster.  @DirtyHairy, this again falls to you, I think? :smiling_imp:

**Comment by ale-79** [comment]
> for me, doing a ROM reload either while in emulation or through the debugger results in the Genesis being detected the second time. 

In my tests, it depends on how I launched the game:

From command line with "-debug" option, reloading from the debugger causes the autodetection to work, while doing it in emulation (which, in this case, opens the debugger again) doesn't.

Without "-debug" option (command line or launcher), autodetection works after reloading (either form debugger or during emulation).

**Comment by sa666666** [comment]
Yes, confirmed for me as well.  In the end, the result is the same, though.  I'm about to do to the 5.0.2 release, so this will have to wait until 5.0.3.

**Comment by DirtyHairy** [comment]
@ale-79 , @sa666666 Thank you both for your analysis :smile: I **think** I know what is going on: in one of my last commits (related to the keyboard controller fixes) I changed the updates of the readout circuit from polling (which was unreliable) to a callback. However, this places new emphasis on the initial state, and I can't see `Genesis.cxx` pull the pin high there :smirk:

However, I cannot reproduce the issue on my system, strange as it seems, for me, both Scramble and Stay Frosty detect the Genesis pad perfectly. I'll do a blind commit to fix this.

**Comment by DirtyHairy** [comment]
Weird, after wiping my stella settings, I can reproduce the issue. Fixed by commit 168c7ba2 .

**Comment by sa666666** [comment]
@thrust26, can you provide a version of Thrust that uses the BoosterGrip controllers?  I want to test a change to Booster class, and it seems the only two ROMs that use it are Omega Race (already tested) and some versions of your Thrust ROM.
